File: viewer/matlab/runme.m
% The following script reads Clarius research data (RF/IQ/B) acquired using
% the Clarius scanner and downloaded from clarius cloud
%
% Note: code is tested on mac, for other operating systems you may need to
% use backslash (\) instead of (/) for relative paths to work. 
%
% Author: Reza Zahiri

close all 
clear all
clc

% add library
addpath('../../common/matlab');

% set data path
path = '../data/wirephantom/';

%% read rf data 
filenameRF = 'phantom_rf.raw';

numFrames = 50;
[dataRF, header] = rdataread([path, filenameRF], numFrames);
if (numFrames > header.frames)
    numFrames = header.frames;
end

% display rf images
figure; 
colormap(gray)
for i = 1:numFrames
    RF = squeeze(dataRF(i,:,:));
    BB = 20*log10(1 + abs(hilbert(RF)));
    subplot(1,2,1), imagesc(RF, [-1000 1000]); title('raw RF')
    subplot(1,2,2), imagesc(BB, [15 70]); title('sample B from RF')
    drawnow;
end

%% read iq data 
filenameIQ = 'phantom_iq.raw';

numFrames = 50;
[dataIQ, header] = rdataread([path, filenameIQ], numFrames);
if (numFrames > header.frames)
    numFrames = header.frames;
end

% separate i and q data
idata = dataIQ(:,1:2:end,:);
qdata = dataIQ(:,2:2:end,:);

% display iq data
figure; 
colormap(gray)
for i = 1:numFrames
    I = squeeze(idata(i,:,:));
    Q = squeeze(qdata(i,:,:));
    BB = 20*log10(1 + sqrt(I.^2 + Q.^2));

    subplot(1,3,1), imagesc(I, [-100 100]); title('I')
    subplot(1,3,2), imagesc(Q, [-100 100]); title('Q')
    subplot(1,3,3), imagesc(BB, [15 70]); title('sample b for IQ')
    drawnow;
end

%% read b or envelope pre scanconversion data 
filenameB = 'phantom_env.raw';

numFrames = 50;
[dataB, header] = rdataread([path, filenameB], numFrames);
if (numFrames > header.frames)
    numFrames = header.frames;
end

% display b/envelope data
figure; 
colormap(gray)
for i = 1:numFrames
    BB = squeeze(dataB(i,:,:));

    imagesc(BB, [0 255]); title('B pre scan')
    drawnow;
end

File: viewer/python/runme.py
"""Read data from the selected folder

This script will load and display Clarius supported file types.

Note: code is tested on mac, for other operating systems you may need to
use backslash (\) instead of (/) for relative paths to work. 

Author: Reza Zahiri

"""

import numpy as np
from scipy.signal import hilbert
import matplotlib.pyplot as plt
import sys
sys.path.append("../../common/python")
import rdataread as rd

if __name__ == '__main__':
    
    # read rf data
    path = "../data/wirephantom/"
    filename = "phantom_rf.raw"
    hdr, timestamps, data = rd.read_rf(path + filename)
    
    # covnert to B 
    numframes = hdr['frames']
    bdata = np.zeros((hdr['lines'], hdr['samples'], hdr['frames']), dtype='float')
    for frame in range(numframes):
        bdata[:,:,frame] = 20 * np.log10( np.abs(1 + hilbert(data[:,:,frame])) )
    
    # display images
    for frame in range(numframes):
        plt.figure(figsize=(10,5))
        plt.subplot(1,2,1)
        plt.imshow(np.transpose(data[:,:,frame]), cmap=plt.cm.gray, aspect='auto', vmin=-1000, vmax=1000 )
        plt.title('RF frame ' + str(frame))
        plt.subplot(1,2,2)
        plt.imshow(np.transpose(bdata[:,:,frame]), cmap=plt.cm.gray, aspect='auto', vmin=15, vmax=70 )
        plt.title('sample B frame ' + str(frame))
        plt.show()    
    
    # read iq data
    path = "../data/wirephantom/"
    filename = "phantom_iq.raw"
    hdr, timestamps, data = rd.read_iq(path + filename)
    # separating i and q data
    idata = data[:,0::2,:]
    qdata = data[:,1::2,:]    
 
    # covnert IQ to B 
    numframes = hdr['frames']
    bdata = np.zeros((hdr['lines'], hdr['samples'], hdr['frames']), dtype='float')
    for frame in range(numframes):
        bdata[:,:,frame] = 10 * np.log10(1 + np.power(idata[:,:,frame], 2) + np.power( qdata[:,:,frame], 2)  )
    
    # display images
    for frame in range(numframes):
        plt.figure(figsize=(15,5))
        plt.subplot(1,3,1)
        plt.imshow(np.transpose(idata[:,:,frame]), cmap=plt.cm.gray, aspect='auto', vmin=-100, vmax=100 )
        plt.title('I frame ' + str(frame))
        plt.subplot(1,3,2)
        plt.imshow(np.transpose(qdata[:,:,frame]), cmap=plt.cm.gray, aspect='auto', vmin=-100, vmax=100 )
        plt.title('Q frame ' + str(frame))
        plt.subplot(1,3,3)
        plt.imshow(np.transpose(bdata[:,:,frame]), cmap=plt.cm.gray, aspect='auto', vmin=15, vmax=70 )
        plt.title('sample B frame ' + str(frame))
        plt.show()
    
    # read b/envelope data
    path = "../data/wirephantom/"
    filename = "phantom_env.raw"
    hdr, timestamps, data = rd.read_env(path + filename)

    # display b data
    numframes = hdr['frames']
    for frame in range(numframes):
        plt.figure(figsize=(5,5))
        plt.imshow(np.transpose(data[:,:,frame]), cmap=plt.cm.gray, aspect='auto', vmin=0, vmax=255 )
        plt.title('envelope frame ' + str(frame))
        plt.show()   
    